### PR TITLE
Varsortability implementation ~~ SortnRegress Algo

### PR DESCRIPTION
**Pull request by Theavinash02**
Added Varsortability along with SortnRegress alogorithm [PR #3247 ]
Issue no : #3392 
Paper: https://arxiv.org/pdf/2102.13647

### CODE REVIEW

**Comment by Theavinash02** [comment]
Hi @ankurankan  ,
Could you please review this PR and kindly share your review comments?
Thanks

**Comment by Theavinash02** [comment]
Hi @ankurankan 
This PR is still in progress. I need to make changes based on our dev meeting. 
I will update and submit for review tomorrow.
Thanks

**Comment by Theavinash02** [comment]
Hi @ankurankan ,
Could you please review this PR and kindly share your review comments?
Thanks

File: pgmpy/causal_discovery/SortnRegress.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    >>> _ = model.fit(data)
+    >>> list(model.causal_graph_.edges())
+    [('X', 'Z'), ('Y', 'Z')]
+
+    References
+    ----------
+    .. [1] Reisach, A. G., Tami, M., Chambaz, A., Seiler, C., & Weichwald, S. (2023).
+       A Scale-Invariant Sorting Criterion to Find a Causal Order in Additive Noise Models.
+       Advances in Neural Information Processing Systems, 36.
+       https://arxiv.org/abs/2303.18211
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: References need to go to `docs/references.bib` file. Then here add `:citep:name`.

Comment (ankurankan) [reply pgmpy/causal_discovery/SortnRegress.py]: Don't use the arxiv version. Find the doi of the paper and then use doi2bib.org to get the bibtex. bibtex goes into references.bib.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    def _fit(self, X):
+        """
+        The fitting procedure for the SortnRegress algorithm.
+
+        Parameters
+        ----------
+        X : pd.DataFrame
+            The data to learn the causal structure from.
+        Returns
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: Add an empty line above.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        -------
+        self : pgmpy.causal_discovery.SortnRegress
+            Returns the instance with the fitted attributes.
+        """
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: We don't need this docstring because this is a private method. The user only calls the `BaseCausalDiscovery.fit` method.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        feature_name_list = list(X.columns)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: Use `self.feature_names_in_` instead of defining your own variable.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        # clone the estimator or use default LinearRegression
+        model_reg = clone(self.estimator) if self.estimator else LinearRegression()
+
+        r2_values = {}
+        for target in feature_name_list:
+            other_nodes = [node for node in feature_name_list if node != target]
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: This is linear time operation. Instead create a set of all nodes outside the `for target in feature_name_list:` loop and do `all_set_nodes - target_node`. This should `O(1)`

File: pgmpy/causal_discovery/SortnRegress.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+            y = X[target].values.astype(float)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: Never implicitly convert the dtype of a variable.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            predictors = X[other_nodes].values.astype(float)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: Same here.

File: pgmpy/causal_discovery/SortnRegress.py (hunk 3 above)
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: Check for `any(X.std() == 0)`. If yes, throw an error saying that the column is constant and ask the user to drop it.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+        model = DAG()
+        model.add_nodes_from(sorted_nodes)
+
+        for i in range(1, len(sorted_nodes)):
+            target = sorted_nodes[i]
+            potential_parents = sorted_nodes[:i]
+
+            y = X[target].values.astype(float)
+            predictors = X[potential_parents].values.astype(float)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: No type conversion

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/causal_discovery/SortnRegress.py (hunk 2 above)
Comment (Theavinash02) [reply pgmpy/causal_discovery/SortnRegress.py]: I have removed the docstring as discussed becasue of private method

File: pgmpy/causal_discovery/SortnRegress.py (hunk 3 above)
Comment (Theavinash02) [reply pgmpy/causal_discovery/SortnRegress.py]: I have removed the docstring as discussed becasue of private method

File: pgmpy/causal_discovery/SortnRegress.py (hunk 6 above)
Comment (Theavinash02) [reply pgmpy/causal_discovery/SortnRegress.py]: removed the type conversion.

File: pgmpy/causal_discovery/SortnRegress.py (hunk 7 above)
Comment (Theavinash02) [reply pgmpy/causal_discovery/SortnRegress.py]: removed the type conversion.

File: pgmpy/causal_discovery/SortnRegress.py (hunk 8 above)
Comment (Theavinash02) [reply pgmpy/causal_discovery/SortnRegress.py]: removed the type conversion.

File: pgmpy/causal_discovery/SortnRegress.py (hunk 3 above)
Comment (Theavinash02) [reply pgmpy/causal_discovery/SortnRegress.py]: Updated the code accordingly

**Comment by Theavinash02** [comment]
Hi @ankurankan 
Could you please review this PR when you are available and kindly share your comments?
Thanks

**Comment by Theavinash02** [comment]
Hi All,
This PR requires further code changes and currently work in progress.
Thanks

File: pgmpy/causal_discovery/SortnRegress.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        self.threshold = threshold
+        self.estimator = estimator
+
+    def _fit(self, X):
+        if any(X.std() == 0):
+            constant_cols = X.columns[X.std() == 0].tolist()
+            raise ValueError(
+                f"The following column(s) have zero variance (constant values): "
+                f"{constant_cols}. Please drop these columns before fitting."
+            )
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py:129]: Could you please create an issue to move this check to `BaseCausalDiscovery`.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        self.feature_names_in_ = list(X.columns)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: Don't assign this; `self.feature_names_in_` is already assigned in `BaseCausalDiscovery.fit` method.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        for target in self.feature_names_in_:
+            other_nodes = list(all_nodes_set - {target})
+
+            y = X[target]
+            predictors = X[other_nodes]
+
+            model_reg.fit(predictors, y)
+            predictions = model_reg.predict(predictors)
+
+            residuals_variance = np.var(y - predictions)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: Can you create an issue to generalize this residual computation for categorical and ordinal variables as well? This would also be applicable to other causal discovery methods such as ANM.

File: pgmpy/causal_discovery/SortnRegress.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        r"""
+        Compute the var-sortability of the input data relative to the graph
+        discovered by this estimator.
+
+        Var-sortability measures how well marginal variances reflect the causal
+        structure. For each directed path in the true DAG, this metric checks whether
+        variance increases monotonically along the path (or remains approximately equal).
+        A score of 1.0 indicates perfect alignment: variances are non-decreasing along
+        all causal paths.
+
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: This is a causal discovery method and should remain in the `causal_discovery` module. Check if this can be a variant of the `SortnRegress` algorithm if there is significant overlap between the two methods.

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/causal_discovery/SortnRegress.py (hunk 10 above)
Comment (Theavinash02) [reply pgmpy/causal_discovery/SortnRegress.py]: Addressed this comment

File: pgmpy/causal_discovery/SortnRegress.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    based on sorting variables by an ordering criterion and iteratively
+    regressing each variable on its predecessors in that order. Two ordering
+    criteria are supported via the ``criterion`` parameter:
+
+    - ``criterion='r2'`` (default): orders variables by ascending global R²,
+      based on the phenomenon that the explainable fraction of a variable's
+      variance, captured by the coefficient of determination (R²), tends to
+      increase along the causal order in linear additive noise models
+      :cite:p:`Reisach2023`.
+    - ``criterion='varsortability'``: orders variables by ascending marginal
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: criterion="variance"

File: pgmpy/causal_discovery/SortnRegress.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+    feature_names_in_: np.ndarray
+        The feature names in the dataset used to learn the causal graph.
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> import numpy as np
+    >>> from pgmpy.causal_discovery import SortnRegress
+    >>> np.random.seed(42)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/SortnRegress.py]: We shouldn't set global seed values. Always use random number generators: https://numpy.org/doc/stable/reference/random/generator.html

**Comment by Theavinash02** [comment]
Hi @ankurankan 
I am sorry for the delay in this PR. I was onsite for the past three weeks. I will complete this PR for this Thursday and start working on the timeseries CD as soon as possible. Thank you for understanding.

**Comment by Theavinash02** [comment]
Hi @ankurankan ,
Could you please re run the checks again?
Thanks

**Comment by Theavinash02** [comment]
Thank you @ankurankan

**Comment by Theavinash02** [comment]
Hi @ankurankan ,
Could you please review this PR and kindly update your review comments ?
Thanks
Avinash

**Comment by Theavinash02** [comment]
Thank you @ankurankan 🤝🏻👍🏻

**Comment by ankurankan** [comment]
@Theavinash02 Please don't add any commits; I am still working on this.

**Comment by Theavinash02** [comment]
Sure, Ankur.👍🏻

**Review by ankurankan — APPROVED** [review]

**Comment by Theavinash02** [comment]
Thank you @ankurankan 
Could you please share an overview of what chnages have been made? 
I will take a look at the code as well :)

**Comment by ankurankan** [comment]
@Theavinash02 The main change has been to separate it out into two classes and remove it from metrics. My understanding of how it can be used as a metric was wrong and doesn't really make sense to do that. Other than that, its just a code refactor with the base class. The idea with the base class is that there are other methods such as LinGAM style methods which follow a similar pattern, and we can implement the shared functionality in the this base class for them.

**Comment by Theavinash02** [comment]
Thank you @ankurankan 👍🏻
